examples/ed25519/src/ed25519.rs

```diff
@@ -0,0 +1,909 @@
+        let a = decompress(pk).ok_or(Error::InvalidPublickey)?;
+        let r_bytes = CompressedEdPoint::from_slice(&signature, 0, 32);
+        let s_bytes = SerializedScalar::from_slice(&signature, 32, 32);
+        if !check_canonical_scalar(s_bytes) {
+            VerifyResult::Err(Error::InvalidS)?;
+        }
+        let r = decompress(r_bytes).ok_or(Error::InvalidR)?;
+        let s = Scalar::from_byte_seq_le(s_bytes);
+        let c = scalar_from_hash(sha512(&r_bytes.concat(&pk).concat(&msg)));
+
+        let z = entropy.slice(16 * i, 16);
+        let z = Scalar::from_byte_seq_le(z.concat(&ByteSeq::new(16)));
+
+        s_sum = s_sum + s * z;
+        r_sum = point_add(r_sum, point_mul(z, r));
+        a_sum = point_add(a_sum, point_mul(z * c, a))
+    }
+
+    let b = decompress(BASE).unwrap();
+    let sb = point_mul(s_sum, b);
+    let check = point_add(point_neg(sb), point_add(r_sum, a_sum));
+    if is_identity(check) {
+        VerifyResult::Ok(())
+    } else {
+        VerifyResult::Err(Error::InvalidSignature)
+    }
+}
+
+/// Batch verification.
+/// Algorithm 3 from <https://eprint.iacr.org/2020/1244.pdf>.
+/// Cofactored verification.
+/// Rejects non-canonical encoding of points.
+/// Rejects small-order points for the public key.
+pub fn alg3_batch_verify(entries: Seq<BatchEntry>, entropy: &ByteSeq) -> VerifyResult {
+    if entropy.len() < 16 * entries.len() {
+        VerifyResult::Err(Error::NotEnoughRandomness)?;
+    }
+    let mut s_sum = Scalar::ZERO();
+    let mut r_sum = point_identity();
+    let mut a_sum = point_identity();
+    for i in 0..entries.len() {
+        let BatchEntry(pk, msg, signature) = entries[i].clone();
+        let a = decompress(pk).ok_or(Error::InvalidPublickey)?;
+        if is_identity(point_mul_by_cofactor(a)) {
+            VerifyResult::Err(Error::SmallOrderPoint)?;
+        }
+        let r_bytes = CompressedEdPoint::from_slice(&signature, 0, 32);
+        let s_bytes = SerializedScalar::from_slice(&signature, 32, 32);
+        if !check_canonical_scalar(s_bytes) {
+            VerifyResult::Err(Error::InvalidS)?;
+        }
+        let r = decompress(r_bytes).ok_or(Error::InvalidR)?;
+        let s = Scalar::from_byte_seq_le(s_bytes);
+        let c = scalar_from_hash(sha512(&r_bytes.concat(&pk).concat(&msg)));
+
+        let z = entropy.slice(16 * i, 16);
+        let z = Scalar::from_byte_seq_le(z.concat(&ByteSeq::new(16)));
+
+        s_sum = s_sum + s * z;
+        r_sum = point_add(r_sum, point_mul(z, r));
+        a_sum = point_add(a_sum, point_mul(z * c, a))
+    }
+
+    let b = decompress(BASE).unwrap();
+    let sb = point_mul(s_sum, b);
+    let check = point_mul_by_cofactor(point_add(point_neg(sb), point_add(r_sum, a_sum)));
+    if is_identity(check) {
+        VerifyResult::Ok(())
+    } else {
+        VerifyResult::Err(Error::InvalidSignature)
+    }
+}
+
+#[cfg(test)]
+extern crate quickcheck;
+
+#[cfg(test)]
+mod test {
```

Comment:
Is there anything here that requires non-public functions? Otherwise it would be nicer to have the tests in the tests folder.